src/plugins/data/server/search/collectors/usage.ts

```diff
@@ -25,34 +25,51 @@ export function usageProvider(core: CoreSetup): SearchUsage {
     return coreStart.savedObjects.createInternalRepository();
   });
 
-  const trackSuccess = async (duration: number, retryCount = 0) => {
-    const repository = await getRepository();
-    try {
-      await repository.incrementCounter(SAVED_OBJECT_ID, SAVED_OBJECT_ID, [
-        { fieldName: 'successCount' },
-        {
-          fieldName: 'totalDuration',
-          incrementBy: duration,
-        },
-      ]);
-    } catch (e) {
-      if (SavedObjectsErrorHelpers.isConflictError(e) && retryCount < MAX_RETRY_COUNT) {
-        setTimeout(() => trackSuccess(duration, retryCount + 1), 1000);
-      }
-    }
+  const collectedUsage: CollectedUsage = {
+    successCount: 0,
+    errorCount: 0,
+    totalDuration: 0,
   };
 
-  const trackError = async (retryCount = 0) => {
-    const repository = await getRepository();
-    try {
-      await repository.incrementCounter(SAVED_OBJECT_ID, SAVED_OBJECT_ID, [
-        { fieldName: 'errorCount' },
-      ]);
-    } catch (e) {
-      if (SavedObjectsErrorHelpers.isConflictError(e) && retryCount < MAX_RETRY_COUNT) {
-        setTimeout(() => trackError(retryCount + 1), 1000);
+  // Instead of updating the search count every time a search completes, we update some in-memory
+  // counts and only update the saved object every ~5 seconds
+  const updateSearchUsage = debounce(
+    async () => {
+      const repository = await getRepository();
+      const counterFields = Object.entries(collectedUsage)
+        .map(([fieldName, incrementBy]) => ({ fieldName, incrementBy }))
+        // Filter out any zero values because `incrementCounter` will still increment them
+        .filter(({ incrementBy }) => incrementBy > 0);
+
+      try {
+        const { attributes } = await repository.incrementCounter<CollectedUsage>(
+          SAVED_OBJECT_ID,
+          SAVED_OBJECT_ID,
+          counterFields
+        );
+
+        // Since search requests may have completed while the saved object was being updated, we minus
+        // what was just updated in the saved object rather than resetting the values to 0
+        collectedUsage.successCount -= attributes.successCount ?? 0;
```

Comment:
You may have a race condition here, if  `collectedUsage` gets updated while the `incrementCounter` request is ongoing.

Reply:
Should be reset from `counterFields` rather than `attributes`